### ISSUE TITLE
Exception handling and error coercion refactor

### ISSUE THREAD

**Issue by kfrz** [department-of-veterans-affairs/va.gov-team#1816]
## User Story or Problem Statement
As a engineer working with `vets-api` I need a maintainable approach for handling errors returned from external (upstream) requests, so that future developers can easily augment and extend custom error handling when necessary.

## Background 

> “Life will not be contained. Life breaks free. It expands to new territories, crashes through barriers, painfully, maybe even dangerously.”

Currently, we approach exception coercion and error reporting in a smattering of various classes and mixins and middlewares. E.g.,: 

- `lib/common/exceptions/external/backend_service_exception.rb`

This file implements a generic error class which is customized with a message and a status code that is manually defined. This status code is ultimately used when logging the error into Rails.logger (or sentry, etc). 

- `config/locales/exceptions.en.yml`
- `lib/common/client/errors.rb`
- `lib/evss/error_middleware.rb`
- `lib/common/client/middleware/response/raise_error.rb`
- `lib/common/exceptions/*/**.rb`

The best approach would be to contain all this related logic in a singular place. 

- It should not be in private methods in the `ApplicationController`, 

### Pain Points

- external errors are not all the same when it comes to how they use status codes, bodies differ, etc.
- exceptions.en.yml is long, hard to parse as a human, have to remember to add to it
- potential (likely duplication) of various status codes. An example is `APPEALSSTATUS50x`, where the status code we provide in the body to the FE is not accurately mapped. 

```yml
      APPEALSSTATUS500:
        <<: *external_defaults
        title: 'Bad Gateway'
        code: 'APPEALSSTATUS500'
        detail: 'Received a 500 response from the upstream server'
        status: 502
      APPEALSSTATUS502:
        <<: *external_defaults
        title: 'Bad Gateway'
        code: 'APPEALSSTATUS502'
        detail: 'Received an invalid response from the upstream server'
        status: 502
```

The historical context here is that the FE doesn't necessarily care about the distinction in terms of user behavior, so this may be a full-stack discussion about how we are using statuses on the Frontend. :question: TODO: follow up? 

- changing a log level can be tricky or hard to discover, also logging method has high complexity rating
- errors are handled in different service classes in different ways (normalize and encapsulate logic)
- distinction between `@key` and `code` in `lib/common/exceptions/external/backend_service_exception.rb` is confusing
- default or "new" errors from upstream services are not really handled gracefully
- `lib/common/client/errors.rb` does this really define all the error types that the client could encounter? why are these types specifically chosen? should we add more? rename?
- Tracing the error/exception handling pathway through repo is a :rabbit: :hole: 
 , partially due to breakers - see `lib/common/client/configuration/base.rb#L92`
- ApplicationController is messy and handles things unrelated to request routing, should be refactored.

## Goal(s)
- Errors are handled by client middleware
- Error handling processors in Sentry manage custom exception coercion

## Objectives or Key Results this is meant to further
- Improve our error handling approach to be consistent between teams

## Acceptance Criteria
- [ ] _What needs to happen or be created?_

**Comment by LindseySaari** [comment]
MVI may be another service similar to EVSS, where we will want to extract some of the error handling logic out

**Comment by omgitsbillryan** [comment]
Some random points:

- The only way to make `vets-api` return an error response code (anything other than `200`) is to raise an exception and let it bubble up to `application_controller.rb`. I find this problematic because there are cases where an upstream service returns a failure that isn't "exceptional", and it would be nice if the controller in question could just `render json: error_json_stuff, status: :not_found`. This is reason for `SKIP_SENTRY_EXCEPTIONS` existing 🤢 .

- The commonality between errors from upstream services is limited. Even 2 different services that are both from EVSS, may have very different error responses. As a result some of our middlewares have become these "if-elsif-elsif-else" messes because they're trying to do too much. Not saying that middlewares aren't useful here, but the current code assumes everything is "common" when in reality each is specific to a given service.

- This could become a big bottomless black hole of work. We probably need some reasonable way to break it up so that so as not to fall into it forever 😨

**Comment by kfrz** [comment]
> - The only way to make vets-api return an error response code (anything other than 200) is to raise an exception and let it bubble up to application_controller.rb. I find this problematic because there are cases where an upstream service returns a failure that isn't "exceptional", and it would be nice if the controller in question could just render json: error_json_stuff, status: :not_found. This is reason for SKIP_SENTRY_EXCEPTIONS existing nauseated_face.

Good point. Do you have suggestions on how this could be a more approachable interface? Just gathering ideas at this point.

> - The commonality between errors from upstream services is limited. Even 2 different services that are both from EVSS, may have very different error responses. As a result some of our middlewares have become these "if-elsif-elsif-else" messes because they're trying to do too much. Not saying that middlewares aren't useful here, but the current code assumes everything is "common" when in reality each is specific to a given service.

- Agreed. Investigating which of the classes are most offensive and will suss out some more solutions in future comments.

> - This could become a big bottomless black hole of work. We probably need some reasonable way to break it up so that so as not to fall into it forever fearful

Agreed. We intend to audit and prioritize the services based on complexity/legacy-ness of code.

**Comment by LindseySaari** [comment]
This is a WIP branch for PPMS refactor (for Sprint 10 tracking purposes): https://github.com/department-of-veterans-affairs/vets-api/tree/1816-ppms-service-refactor
